File: templates/arrays/two_pointers.go
package arrays

func TwoSumSorted(nums []int, target int) (int, int, bool) {
	left, right := 0, len(nums)-1
	for left < right {
		sum := nums[left] + nums[right]
		switch {
		case sum == target:
			return left, right, true
		case sum < target:
			left++
		default:
			right--
		}
	}
	return -1, -1, false
}

func RemoveDuplicatesSorted(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	write := 1
	for read := 1; read < len(nums); read++ {
		if nums[read] != nums[read-1] {
			nums[write] = nums[read]
			write++
		}
	}
	return write
}

func MaxArea(height []int) int {
	left, right := 0, len(height)-1
	best := 0

	for left < right {
		width := right - left
		h := height[left]
		if height[right] < h {
			h = height[right]
		}
		area := width * h
		if area > best {
			best = area
		}

		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return best
}

File: templates/arrays/two_pointers.py
def two_sum_sorted(nums: list[int], target: int) -> tuple[int, int, bool]:
    left, right = 0, len(nums) - 1
    while left < right:
        current = nums[left] + nums[right]
        if current == target:
            return left, right, True
        if current < target:
            left += 1
        else:
            right -= 1
    return -1, -1, False


def remove_duplicates_sorted(nums: list[int]) -> int:
    if not nums:
        return 0

    write = 1
    for read in range(1, len(nums)):
        if nums[read] != nums[read - 1]:
            nums[write] = nums[read]
            write += 1
    return write


def max_area(height: list[int]) -> int:
    left, right = 0, len(height) - 1
    best = 0

    while left < right:
        width = right - left
        area = width * min(height[left], height[right])
        best = max(best, area)

        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return best
